### ISSUE TITLE
VRRP traffic dropped by Cilium - CT: Unknown L4 protocol

### ISSUE THREAD

**Issue by sbocahu** [cilium/cilium#31502]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

VRRP traffic is dropped by Cilium with 1.15.2. was working with 1.14.2

keepalived + haproxy on 3 nodes to setup HA apiserver, upgraded cilium from 1.14.2 to 1.15.2; keepalived can't see any peer on each server and hubble reports dropped traffic: `CT: Unknown L4 protocol`

 tried without and with hostFirewall enabled after having found https://github.com/cilium/cilium/issues/18347

### Cilium Version

cilium-cli: v0.14.6 compiled with go1.20.4 on linux/amd64
cilium image (default): v1.13.3
cilium image (stable): v1.15.1
cilium image (running): v1.15.2

### Kernel Version

Linux control-001 6.1.0-18-amd64 #1 SMP PREEMPT_DYNAMIC Debian 6.1.76-1 (2024-02-01) x86_64 GNU/Linux

### Kubernetes Version

Client Version: version.Info{Major:"1", Minor:"27", GitVersion:"v1.27.4", GitCommit:"fa3d7990104d7c1f16943a67f11b154b71f6a132", GitTreeState:"clean", BuildDate:"2023-07-19T12:20:54Z", GoVersion:"go1.20.6", Compiler:"gc", Platform:"linux/amd64"}
Kustomize Version: v5.0.1
Server Version: version.Info{Major:"1", Minor:"27", GitVersion:"v1.27.12", GitCommit:"12031002905c0410706974560cbdf2dad9278919", GitTreeState:"clean", BuildDate:"2024-03-15T02:06:14Z", GoVersion:"go1.21.8", Compiler:"gc", Platform:"linux/amd64"}

### Relevant log output

```shell
hubble observe --type=drop -f
Mar 19 13:34:43.419: 172.16.41.2 (host) <> 172.16.41.3 (kube-apiserver) CT: Unknown L4 protocol DROPPED (IPv4)
Mar 19 13:34:43.445: 172.16.41.1 (host) <> 172.16.41.2 (kube-apiserver) CT: Unknown L4 protocol DROPPED (IPv4)
Mar 19 13:34:43.445: 172.16.41.1 (host) <> 172.16.41.3 (kube-apiserver) CT: Unknown L4 protocol DROPPED (IPv4)
Mar 19 13:34:43.592: 172.16.41.3 (host) <> 172.16.41.1 (kube-apiserver) CT: Unknown L4 protocol DROPPED (IPv4)
Mar 19 13:34:43.592: 172.16.41.3 (host) <> 172.16.41.2 (kube-apiserver) CT: Unknown L4 protocol DROPPED (IPv4)
Mar 19 13:34:44.419: 172.16.41.2 (host) <> 172.16.41.1 (kube-apiserver) CT: Unknown L4 protocol DROPPED (IPv4)
Mar 19 13:34:44.419: 172.16.41.2 (host) <> 172.16.41.3 (kube-apiserver) CT: Unknown L4 protocol DROPPED (IPv4)
Mar 19 13:34:44.445: 172.16.41.1 (host) <> 172.16.41.2 (kube-apiserver) CT: Unknown L4 protocol DROPPED (IPv4)
Mar 19 13:34:44.445: 172.16.41.1 (host) <> 172.16.41.3 (kube-apiserver) CT: Unknown L4 protocol DROPPED (IPv4)
```

### Cilium Users Document

- [X] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by aditighag** [comment]
This looks like a dup of https://github.com/cilium/cilium/issues/18347? Could you follow-up on the same issue?

**Comment by sbocahu** [comment]
Definitely related but not dup of the initial issue. I propose to retitle the issue; hostfirewall enabled/disabled has no effect on this issue in 1.15 and other protocols are impacted as well. seem to be a regression as it was working in my setup in 1.14

**Comment by kvaster** [comment]
I just can confirm that this are two different bugs. One is related to host firewall (#18347) and this one is related to bpf masquerade somehow.

What can I observe:

* use keepalived with LAN ip4 - it's all OK
* use keepalived with WAN ip4 - I can see 2-4 vrrp packets after keepalived start, depending on nodes/ips and after that packets are dropped. After keepalived restart I can see again 2-4 vrrp packets.

This is all related to setup without host firewall. I'm preparing PR right now for the fix, but don't know if it's right way (I'm not cilium maintainer).

**Comment by kvaster** [comment]
@sbocahu, can you please tell what other protocols are impacted as well ? Is this also a multicast traffic ?

**Comment by sbocahu** [comment]
@kvaster OSPF is reported to be impacted as well: https://github.com/cilium/cilium/issues/18347#issuecomment-1752492445

**Comment by kvaster** [comment]
OSPF is also based on multicast. But I should say that my fix is only for host multicast traffic without hostfirewall enabled.

**Comment by sbocahu** [comment]
I'm considering using hostFirewall in the future however for the time being I'ld be glad to upgrade to 1.15 so thank you very much @kvaster. Hope a new minor fixing this soon :) !

**Comment by julianwiedmann** [comment]
I *think* I see what's up. This might have been introduced by https://github.com/cilium/cilium/pull/26501.  Can you attach your sysdump / cilium config to confirm whether that matches up?

Would you be able to reproduce on latest `main`, and potentially test a PR?

**Comment by julianwiedmann** [comment]
https://github.com/cilium/cilium/pull/31820 is my current thinking. But I wouldn't expect it to help with HostFW enabled ...

**Comment by kvaster** [comment]
@julianwiedmann , I can test on latest main without any problm. Will do it tomorrow and will get back. Also testing PR is not a problem aswell.

**Comment by kvaster** [comment]
@julianwiedmann , I've made a check. The problem is gone with latest `main`.

**Comment by julianwiedmann** [comment]
> @julianwiedmann , I've made a check. The problem is gone with latest `main`.

excellent, thank you!

**Comment by sbocahu** [comment]
This issue has been fixed and integrated into 1.15.3 or 1.15.4
thanks !
